File: 1582-design-browser-history/design-browser-history.py
class BrowserHistory:

    def __init__(self, homepage: str):
        self.arr=[homepage]
        self.index=0

        

    def visit(self, url: str) -> None:
        self.index+=1
        self.arr=self.arr[:self.index]
        self.arr.append(url)

    def back(self, steps: int) -> str:
        self.index=max(0,self.index-steps)
        return self.arr[self.index]
    def forward(self, steps: int) -> str:
        self.index=min(len(self.arr)-1,self.index+steps)
        return self.arr[self.index]

        


# Your BrowserHistory object will be instantiated and called as such:
# obj = BrowserHistory(homepage)
# obj.visit(url)
# param_2 = obj.back(steps)
# param_3 = obj.forward(steps)

File: 1582-design-browser-history/design-browser-history.java
class BrowserHistory {
    List<String> list;
    int size;
    int index;

    public BrowserHistory(String homepage) {
        this.list= new LinkedList<>();
        list.add(homepage);
        this.size=1;
        this.index=0;
    }
    
    public void visit(String url) {
        index++;
        if(index==list.size()){
            list.add(url);
        }else{
            list.set(index,url);
        }
        this.size=index+1;
    }
    
    public String back(int steps) {
        this.index= Math.max(0,this.index-steps);
        return list.get(index);
    }
    
    public String forward(int steps) {
        this.index=Math.min(this.size-1, this.index+steps);
        return list.get(index);
        
    }
}

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory obj = new BrowserHistory(homepage);
 * obj.visit(url);
 * String param_2 = obj.back(steps);
 * String param_3 = obj.forward(steps);
 */